m4/dovecot.m4

```diff
@@ -342,6 +342,19 @@ AC_DEFUN([DC_DOVECOT_HARDENING],[
 	AC_LD_RELRO
 ])
 
+AC_DEFUN([DC_DOVECOT_FUZZER],[
+        AC_ARG_ENABLE(fuzzer,
+        AS_HELP_STRING([--enable-fuzzer=yes], [Enable clang fuzzer (default: no)]),
+                enable_fuzzer=$enableval,
+                enable_fuzzer=yes)
+	AS_IF([test x$enable_fuzzer = xyes], [
+		CFLAGS="$CFLAGS -fsanitize=fuzzer-no-link"
+		FUZZER_LDFLAGS=-fsanitize=fuzzer
```

Comment:
This should be made more generic
oss-fuzz uses environment variable `LIB_FUZZING_ENGINE`
On MacOS, this does not work and you have to set `LIB_FUZZING_ENGINE=/path/to/libFuzzer.a`

Reply:
It's not very clear to me how these are supposed to work. So clang has libfuzzer and I got the above parameters from it. But then oss-fuzz does something else on top of clang libfuzzer? Or not? Don't I need any of the -fsanitize=fuzzer flags to run with oss-fuzz after all? I don't really understand why clang wants to build sources with this -fsanitize=fuzzer anyway.

Reply:
You can see it this way : 
There are two things :
- fuzz targets : those are the files like fuzz-imap-utf7.c with use of `LLVMFuzzerTestOneInput`
- fuzz drivers : this can be libFuzzer (from clang with `-fsanitize=fuzzer`), but there are others such as AFL... Oss-fuzz uses libFuzzer, AFL and HongFuzz as drivers. That is why they use this generic `LIB_FUZZING_ENGINE`. Basically they are a `main` function calling repeatedly `LLVMFuzzerTestOneInput`...

> Don't I need any of the -fsanitize=fuzzer flags to run with oss-fuzz after all?

These are set by `CFLAGS` in oss-fuzz

Reply:
@sirainen  in my PR to OSS-Fuzz where I added the Dovecot project you can see how I only use the CFLAGS and LIB_FUZZING_ENGINE when building the fuzzers, but never actually set variables as OSS-Fuzz does this for you for free https://github.com/google/oss-fuzz/pull/3502/commits/91e5c2eaedf354fb640961dbe96a15286d793b58#diff-422c46e236d34703c859d7d6378624e8

Reply:
OK, so with oss-fuzz you don't need to set any special CFLAGS? I tested with clang that it really does need them or it'll fail with:

ERROR: no interesting inputs were found. Is the code instrumented for coverage? Exiting.

I've now implemented `--enable-fuzzer=clang` for the clang method and `--enable-fuzzer=oss` that only links to LIB_FUZZING_ENGINE. Does this work?

Reply:
That's correct - the default values that OSS-Fuzz sets the flags to is something like this: `CFLAGS= -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link` 
and `CXXFLAGS=-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -stdlib=libc++`

Reply:
The error you get is because the "fuzzer-no-link" was not used when compiling the object files of the projekt. In general, we should use "-fsanitize=fuzzer-no-link" when compiling and "-fsanitize=fuzzer" when linking

Reply:
Ah, so CFLAGS are set in environment so it overrides the configure-provided CFLAGS. What does LIB_FUZZING_ENGINE expand to then? Does it contain the -fsanitize=fuzzer when linking, or is that added to LDFLAGS environment also, or something?

Reply:
Yeah, LIB_FUZZING_ENGINE expands to -fsanitize=fuzzer. However, notice this is the default setting and Oss-Fuzz might set the flags a bit different if, for example, we are compiling for a "coverage" build. Then, CFALGS expands to something like `CFLAGS=-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION  -fprofile-instr-generate -fcoverage-mapping -pthread -Wl,--no-as-needed -Wl,-ldl -Wl,-lm -Wno-unused-command-line-argument` and `CXXFLAGS=-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION  -fprofile-instr-generate -fcoverage-mapping -pthread -Wl,--no-as-needed -Wl,-ldl -Wl,-lm -Wno-unused-command-line-argument -stdlib=libc++`

Reply:
Okay, so another attempt: Now there's again only `configure --enable-fuzzer` which sets `CFLAGS=-fsanitize=fuzzer-no-link` but that doesn't matter because oss-fuzz overwrites it anyway. It also uses `-fsanitize=fuzzer` for linking, except if `$LIB_FUZZING_ENGINE` environment is set then that is used for linking instead.

Reply:
Sounds good to me!

Reply:
> Ah, so CFLAGS are set in environment so it overrides the configure-provided CFLAGS.

CFLAGS are indeed set in environment and configure should only _expand_ them (there should be no overriding)